src/main/java/com/vlkan/pubsub/PubsubClient.java

```diff
@@ -0,0 +1,274 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permits and
+ * limitations under the License.
+ */
+
+package com.vlkan.pubsub;
+
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.vlkan.pubsub.model.*;
+import com.vlkan.pubsub.util.MicrometerHelpers;
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Timer;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpHeaderValues;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import reactor.core.publisher.Mono;
+import reactor.netty.http.client.HttpClient;
+
+import javax.annotation.Nullable;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Map;
+import java.util.Objects;
+import java.util.WeakHashMap;
+import java.util.function.Supplier;
+
+public class PubsubClient {
+
+    public static final ObjectMapper DEFAULT_OBJECT_MAPPER = new ObjectMapper()
+            // To allow backward-compatible future enhancements in the
+            // protocol, disable failure on unknown properties.
+            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+
+    private final PubsubClientConfig config;
+
+    private final ObjectMapper objectMapper;
+
+    private final PubsubAccessTokenCache accessTokenCache;
+
+    private final HttpClient httpClient;
+
+    private final MeterRegistry meterRegistry;
+
+    private final Map<String, Timer> timerByRequestUrl = new WeakHashMap<>();
+
+    private final Map<String, Counter> counterByRequestUrl = new WeakHashMap<>();
+
+    private PubsubClient(Builder builder) {
+        this.config = builder.config;
+        this.objectMapper = builder.objectMapper;
+        this.accessTokenCache = builder.accessTokenCache;
+        this.httpClient = builder.httpClient;
+        this.meterRegistry = builder.meterRegistry;
+    }
+
+    Mono<PubsubPullResponse> pull(String projectName, String subscriptionName, PubsubPullRequest pullRequest) {
+        String requestUrl = String.format(
+                "%s/v1/projects/%s/subscriptions/%s:pull",
+                config.getBaseUrl(), projectName, subscriptionName);
+        Supplier<String[]> tagSupplier = () -> new String[]{
+                "operation", "pull",
+                "projectName", projectName,
+                "subscriptionName", subscriptionName
+        };
+        return executeRequest(requestUrl, pullRequest, PubsubPullResponse.class, config.getPullTimeout())
+                .transform(mono -> MicrometerHelpers.measureLatency(
```

Comment:
idea: `MicrometerHelpers.measureLatency` could return a higher order `Function<Mono, Mono>` instead of accepting Mono as the last parameter, so that you can store it in the field and apply with `.transform(this.measureLatencyFn)`

Reply:
Sounds like a great idea! Let me try to work it out.

Reply:
Hrm... Noticed a show stopper: The `projectName`, `subscriptionName`, and `topicName` tags are populated from the method parameters. Hence I need to create them on a on-call basis -- unless they already exist in the `timerByRequestUrl` and `counterByRequestUrl` caches.

Reply:
The beauty of FP is that you can have many orders of functions:
```java
.transform(this.measureLatencyFn.apply(tagSupplier)
```

:) Or, if you want to keep it simple, you still can have:
```java
.transform(MicrometerHelpers.measureLatency(
    meterRegistry,
    timerByRequestUrl,
    requestUrl,
    tagSupplier
))
```